### ISSUE TITLE
Interpretation of nominal srate

### ISSUE THREAD

**Issue by sjeung** [sccn/xdf#56]
Hello everyone, 

I have a couple of questions about some metadata fields that have been a bit confusing for me. 
It would be really awesome if someone could clarify these points so we know how to implement experiments in the future :)

1. How is nominal sampling rate of 0 to be interpreted? 
Is this, a) necessarily a stream with no nominal srate in any sense, such as event streams OR b) can be a continuous stream with irregular sampling intervals? 

In the [specification](https://github.com/sccn/xdf/wiki/Specifications), I read "If the stream has an irregular sampling rate (that is, the samples are not spaced evenly in time, for example in an event stream), this value must be 0." 
Now, depending on what "irregular sample rate" means, nominal srate of 0 may also be assigned to a motion capture system of 90 Hz desirable (or nominal?) sampling rate that has some jitters in time stamps. [LSL for unity](https://github.com/labstreaminglayer/LSL4Unity) appears to follow this interpretation and uses this flag to decide whether time stamps generated are perfectly regular or not. 

On the other hand, [xdf2fieldtrip](https://github.com/fieldtrip/fieldtrip/blob/master/xdf2fieldtrip.m) checks nominal srate to determine whether a stream is continuous. In fact, in the group I am working in, there are data sets created using the same hardware but use either nominal srate of 0 or 90 for continuous streams depending on LSL4Unity configuration, so this has been a bit tricky to deal with.  

The practical question is, is it OK to use nominal srate of 0 when there actually is some kind of nominal srate (90 Hz)? The presence of nominal srate appears to be mixed up with regularity of sample intervals or continuousness of streams, which I think all mean different things. Can we think of adding new fields for these properties?

2. How are the quaternion orientation units to be interpreted in motion capture metadata? 

Thanks a lot in advance!

**Comment by chkothe** [comment]
Hi, here's an attempt at a clarification for question (1): 

Many devices (and some software applications) are deliberately designed so as to maintain (within some tolerance) a specified regular rate at which values are measured (sampled). When that is the case, one should consider declaring the specified rate as the nominal rate. When that is **not** the case, the stream should be declared irregular (that is, nominal_rate=0). There is a grey zone when a device exhibits deviations from some intended or specified rate that are sufficiently large (*) that it is better treated as irregular. So this means that nominal_rate=0 just means that one cannot expect there to be (even approximate) regularity. This includes events streams, but that's only a subset.

*: The remaining question is what's "sufficiently large" for a stream to qualify as irregular. Almost any real-world device or system is subject to some pervasive deviations that we'd clearly consider benign or acceptable:

1. no real-world system (short of an atomic clock) maintains the nominal rate anywhere near "perfectly" -- e.g. technical [limitations](https://en.wikipedia.org/wiki/Crystal_oscillator#Stability) leave even most professional-grade devices with as much as +/- 0.01% of error between the nominal (specified) rate and effective (measured) sampling rate. Devices or streams clocked off the mains frequency (or, e.g., monitor refresh rate) could have as much as [1% error](https://www.google.com/search?q=mains+frequency+accuracy). Furthermore, those deviations even change on the order of minutes to hours for various reasons (e.g. temperature), so the sampling rate is not even stable.
2. almost any real-world system can temporarily cease to send data (e.g. run out of battery, crash, lose connectivity), which results in occasional (usually lengthy) gaps in the time series. It helps that sufficiently long gaps are quite easy to detect if one knows the nominal rate.

Among the less benign deviations are the following:

3. a device that sporadically drops samples (e.g., certain cameras): these issues are extremely hard to detect -- however, in those cases more metadata is actually helpful, i.e., those devices should ideally declare both the nominal sampling rate and also a (recently-proposed, not-yet-documented) flag that indicates that the stream can drop frames (`synchronization/can_drop_samples`).
4. a device whose sampling interval may be steady at times but can become quite erratic (e.g., as in video game frame rates). We currently have no flag "can_be_erratic" or the like (let alone applications using it), so we have to assume that an application reading such a stream would be unable to detect the erratic periods (they wouldn't register as gaps) and that they would almost certainly generate bogus outputs when treating that as regular. It's also not clear that a flag would really help, or whether there should be another kind of sampling rate like the "desired_rate" or the like, which describes the behavior on a lucky day (and which can be used to detect those lucky stretches for the few (probably fairly rare) applications that want to do that).

The above should also make it clear that, just because a stream had at some point been measured to have an average sampling rate of say 53.6 Hz, that's no reason (at all) to declare that as the nominal sampling rate -- the nominal sampling rate will almost always be explicitly built into some sort of engineered process for maintaining that rate (even if that process isn't perfect or always successful).

Lastly, it's also important to note that the time stamps themselves are measurements, and especially for hardware streams they can be jittery even though the underlying sampling process is (usually) far more regular. The regularity statement is about the sampling process (that generates the data values), not the time stamps. The time stamp jitter would usually be addressed in many applications by dejittering or linearizing the time stamps (likely after detecting and respecting any data gaps) to get improved estimates of the actual time stamps.

**Comment by cboulay** [comment]
I was about to send my response as Christian's popped up. Some of this may be redundant with what he said but there's a useful example in there to help you think about it.

First, the idea of "continuous" is more of a scientific thing than an engineering thing. Indeed every data source in LSL is probably "continuous", depending on how you define it, otherwise you wouldn't need something like LSL. The use of 'nominal_rate' is orthogonal to the question of whether or not a stream is continuous. They don't mean the same thing, and "continuous" or not doesn't have a good definition (within the context of all LSL streams) so I don't think it merits a new flag.

The way I think about it: whether or not I set an outlet to have a non-zero nominal_rate depends on whether or not I want the XDF importers or the LSL postprocessors (if invoked by the Inlet) to try to smooth out the timestamps.

Let's say you're using a wiimote. When you are actively shaking it, its data rate is fairly consistent. However, when you set it down it ceases to transmit, so its data rate drops to 0. Now you start shaking it again, and the period at rest was less than the [`jitter_break_threshold_seconds`](https://github.com/xdf-modules/pyxdf/blob/master/pyxdf/pyxdf.py#L77-L78). What do you want to happen? Well if you specified a non-zero nominal rate, then the xdf importer will assume that the first of the new timestamps should be 1/srate after the last received timestamp and will attempt to smooth out the discrepancies. But this is incorrect. To avoid this we could set the wiimote outlet to have a nominal_rate of 0 so all timestamps are taken at face value. This makes it impossible for the outlet to use `push_chunk` and for any receivers to do filtering or other signal processing because they don't know the expected rate. There may be other solutions to this particular problem, but I'm just using it to illustrate.

>  LSL for unity appears to follow this interpretation and uses this flag to decide whether time stamps generated are perfectly regular or not.

LSL4Unity is not a ground truth for anything. I'm sure Markus would agree. 

Different software might use nominal_rate as a shortcut to say "if it's nominal_rate is 0 then this stream contains events". I usually take it a couple steps further and only assume that a stream is an event stream if its type is 'Markers', it has a single channel, and the dtype is strings.

@sjeung - please open a separate issue for the question about Mocap specifications.

**Comment by chkothe** [comment]
Regarding question (2):

(edit: moved over to issue https://github.com/sccn/xdf/issues/57)

**Comment by sjeung** [comment]
Thanks a lot @chkothe and @cboulay for your prompt and detailed answers. Both answers help me get a much clearer picture. My group is indeed dealing with the less benign type of deviations (from a VR setup) that Christian described. Both nominal rate = 0 and nominal rate = 90 + can drop samples flag 'on' seem to work fairly well when we use load_xdf with default options but I will try generating and comparing data from the two configurations in the next days. I also agree with Chadwick that nominal srate is not the best way to separate different stream types. 

I have created a separate issue #57 for mocap metadata, as suggested.
